src/rosdiscover/observer/ros1.py

```diff
@@ -0,0 +1,261 @@
+_CANCEL = re.compile(r'(.*)/cancel')
+_STATUS = re.compile(r'(.*)/status')
+_FEEDBACK = re.compile(r'(.*)/feedback')
+_FEEDBACK_FMT = re.compile(r'(.*)Feedback')
+_RESULT = re.compile(r'(.*)/result')
+_RESULT_FMT = re.compile(r'(.*)Result')
+
+
+@attr.s(slots=True, auto_attribs=True)
+class _Action:
+    """Class for action information."""
+    name: str
+    fmt: str
+
+
+@attr.s(slots=True, auto_attribs=True)
+class _Node:
+    """Class for partial node information."""
+    name: str
+    publishers: Set[str] = attr.ib(default=set())
+    subscribers: Set[str] = attr.ib(default=set())
+    provides: Set[str] = attr.ib(default=set())
+
+    @staticmethod
+    def filter_topics_for_action(ros: ROS1, goal_related_topics: Set[str],
+                                 result_related_topics: Set[str]) -> Collection[_Action]:
+        """
+        Filter the topics in :code:`goal_related_topics` and `:code:result_related_topics` to
+        pull out action-related topics.
+
+        Parameters
+        ----------
+        ros: ROS1
+            Information about the state of the ROS system.
+        goal_related_topics: Set[str]
+            The topics where the action goal is meant to be. Servers will subscribe to this,
+            clients will publish.
+        result_related_topics: Set[str]
+            The topics where the action result should be. Servers will publish, clients will
+            subscribe.
+
+        Returns
+        -------
+        A collection of actions that were removed from the topic collections.
+        """
+        actions = []
+        # Copy the topic set containing the goal because we are going to mutate it
+        topic_copy = set(goal_related_topics)
+
+        # Process the nodes for topics that are action related
+        for topic in topic_copy:
+            goal_match = _GOAL.match(topic)
+            if goal_match:  # The topic might be a action goal
+                fmt_match = _GOAL_FMT.match(ros.topic_to_type[topic])
+                if fmt_match:  # The topic is an action goal
+                    # Have the right goal and format matches. Check if other topics are there
+                    action = goal_match.group(1)
+                    fmt = fmt_match.group(1)
+                    if _Node.has_all_action_topics(action, fmt,
+                                                   ros, goal_related_topics, result_related_topics):
+                        # Remove the topics from the right collections and replace as an action
+                        goal_related_topics.remove(f"{action}/goal")
+                        goal_related_topics.remove(f"{action}/cancel")
+                        result_related_topics.remove(f"{action}/status")
+                        result_related_topics.remove(f"{action}/feedback")
+                        result_related_topics.remove(f"{action}/result")
+                        actions.append(_Action(action, fmt))
+        return actions
+
+    @staticmethod
```

Comment:
You can use `@classmethod` to avoid the need to repeat `_Node` in the static method calls (e.g., `_Node.has_cancel(...)` becomes `cls.has_cancel(...)`.

Reply:
Done